File: hdv-foundation/personamatrix/scoring.py
"""scoring.py -- Python twin of knoll/scoring.ts (behavioral anomaly scoring).

A stdlib-only mirror of the TypeScript behavioral scorer, for Colab experimentation and
offline tuning of KNOLL's additive scoring gate. It is ADDITIVE to the six virtual laws;
it never replaces them. numpy is optional -- everything here uses only the standard library.

ENFORCEMENT THRESHOLD (KNOLL active-router change): the default deny threshold is 0.34
(34%), mirroring knoll/scoring.ts. Lowered from the legacy 0.6 when KNOLL became an
INDEPENDENT ACTIVE ROUTER: at or above 34% the TypeScript KNOLL denies the packet AND trips
an immediate system-level freeze + packet quarantine (see knoll/freeze.ts). This Python twin
keeps the same threshold so offline tuning matches live enforcement. The Shannon-entropy
feature (`intent_entropy`) is one contributor that can push a high-entropy blob over 34%.

Features (each normalized 0..1, higher = more suspicious):
    rate             -- recent request volume from the source (flooding)
    intent_entropy   -- character entropy of intent + string payload (random blobs)
    malicious_hits   -- soft suspicious-keyword hits (weaker than the hard LAW 6 block)
    endpoint_risk    -- inherent risk of the (source -> destination) pair
    payload_size     -- normalized serialized payload size
    priority_abuse   -- CRITICAL priority used where it shouldn't be (queue-jumping)
    source_reputation-- accumulated risk history for the source
"""
from __future__ import annotations

import json
import math
import re
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

# Soft suspicious keywords -- intentionally weaker than the hard-blocking patterns.
SOFT_SUSPICIOUS = [
    re.compile(p, re.IGNORECASE)
    for p in (
        r"\bpassword\b", r"\bcredential", r"\btoken\b", r"\bsecret", r"\bsudo\b",
        r"\broot\b", r"\badmin\b", r"\boverride\b", r"\bbypass\b", r"\bexploit\b",
        r"\bescalat", r"\bbase64\b", r"\bpayload\b", r"\bbackdoor\b",
    )
]

ENDPOINT_RISK = {
    "APEX->VISION": 0.6, "APEX->DREAM": 0.2, "APEX->HOPE": 0.1, "HOPE->APEX": 0.1,
    "DREAM->HOPE": 0.15, "VISION->HOPE": 0.2, "DREAM->APEX": 0.15, "VISION->APEX": 0.2,
    "DREAM->VISION": 1.0, "VISION->DREAM": 1.0,
}

# Weights sum to 1.0. `intent_entropy` was raised (0.10 -> 0.20, pulling from the low-signal
# `rate` and `payload_size` features) so a Shannon-entropy SPIKE can meaningfully contribute to
# crossing the 34% deny threshold. `malicious_hits` stays the dominant hard signal. Mirrors
# knoll/scoring.ts DEFAULT_WEIGHTS.
DEFAULT_WEIGHTS: Dict[str, float] = {
    "rate": 0.10,
    "intent_entropy": 0.20,
    "malicious_hits": 0.30,
    "endpoint_risk": 0.15,
    "payload_size": 0.05,
    "priority_abuse": 0.10,
    "source_reputation": 0.10,
}


def _clamp01(x: float) -> float:
    return max(0.0, min(1.0, x))


def _collect_strings(value: Any) -> List[str]:
    if isinstance(value, str):
        return [value]
    if isinstance(value, list):
        out: List[str] = []
        for v in value:
            out.extend(_collect_strings(v))
        return out
    if isinstance(value, dict):
        out = []
        for v in value.values():
            out.extend(_collect_strings(v))
        return out
    return []


def normalized_entropy(text: str) -> float:
    """Shannon character entropy normalized against a 6-bit ceiling."""
    if not text:
        return 0.0
    freq: Dict[str, int] = {}
    for ch in text:
        freq[ch] = freq.get(ch, 0) + 1
    n = len(text)
    h = 0.0
    for c in freq.values():
        p = c / n
        h -= p * math.log2(p)
    return _clamp01(h / 6.0)


def _soft_hits(text: str) -> float:
    hits = sum(1 for pat in SOFT_SUSPICIOUS if pat.search(text))
    return _clamp01(hits / 3.0)


def _payload_size(data: Any) -> float:
    return _clamp01(len(json.dumps(data, sort_keys=True)) / 8192.0)


def _priority_abuse(packet: Dict[str, Any]) -> float:
    priority = packet.get("priority", "STANDARD")
    if priority != "CRITICAL":
        return 0.0 if priority == "BACKGROUND" else 0.1
    intent = str(packet.get("intent", "")).lower()
    trivial = re.search(r"\b(query|what|who|when|explain|describe|hello|ping)\b", intent) is not None
    return 1.0 if trivial else 0.6


def _endpoint_risk(source: str, destination: str) -> float:
    return ENDPOINT_RISK.get(f"{source}->{destination}", 0.25)


def extract_features(packet: Dict[str, Any], recent_count: int, rate_soft_cap: int, reputation: float) -> Dict[str, float]:
    haystack = " \n ".join([str(packet.get("intent", ""))] + _collect_strings(packet.get("data", {})))
    return {
        "rate": _clamp01(recent_count / max(1, rate_soft_cap)),
        "intent_entropy": normalized_entropy(haystack),
        "malicious_hits": _soft_hits(haystack),
        "endpoint_risk": _endpoint_risk(packet.get("source", ""), packet.get("destination", "")),
        "payload_size": _payload_size(packet.get("data", {})),
        "priority_abuse": _priority_abuse(packet),
        "source_reputation": _clamp01(reputation),
    }


@dataclass
class BehavioralScore:
    score: float
    threshold: float
    is_anomalous: bool
    flagged: bool
    features: Dict[str, float]
    contributions: Dict[str, float]


@dataclass
class BehavioralScorer:
    """Stateful anomaly scorer mirroring knoll/scoring.ts."""

    # 0.34 (34%) deny threshold -- KNOLL active-router enforcement (was 0.6). See module docstring.
    threshold: float = 0.34
    flag_threshold: float = 0.2
    rate_window_s: float = 1.0
    rate_soft_cap: int = 20
    weights: Dict[str, float] = field(default_factory=lambda: dict(DEFAULT_WEIGHTS))
    _windows: Dict[str, List[float]] = field(default_factory=dict)
    _reputation: Dict[str, float] = field(default_factory=dict)

    def score(self, packet: Dict[str, Any], now: Optional[float] = None) -> BehavioralScore:
        source = str(packet.get("source", "UNKNOWN"))
        recent = self._observe_rate(source, now if now is not None else time.time())
        reputation = self._reputation.get(source, 0.0)

        feats = extract_features(packet, recent, self.rate_soft_cap, reputation)
        contributions: Dict[str, float] = {}
        total = 0.0
        for key, weight in self.weights.items():
            c = feats[key] * weight
            contributions[key] = round(c, 4)
            total += c
        total = _clamp01(total)

        is_anom = total >= self.threshold
        flagged = (not is_anom) and total >= self.flag_threshold

        if is_anom:
            self._bump(source, 0.25)
        elif flagged:
            self._bump(source, 0.05)
        else:
            self._bump(source, -0.02)

        return BehavioralScore(
            score=round(total, 4),
            threshold=self.threshold,
            is_anomalous=is_anom,
            flagged=flagged,
            features=feats,
            contributions=contributions,
        )

    def reputation_of(self, source: str) -> float:
        return round(self._reputation.get(source, 0.0), 4)

    def reset(self) -> None:
        self._windows.clear()
        self._reputation.clear()

    def _observe_rate(self, source: str, now: float) -> int:
        cutoff = now - self.rate_window_s
        arr = [t for t in self._windows.get(source, []) if t > cutoff]
        arr.append(now)
        self._windows[source] = arr
        return len(arr)

    def _bump(self, source: str, delta: float) -> None:
        self._reputation[source] = _clamp01(self._reputation.get(source, 0.0) + delta)

File: hdv-foundation/knoll/scoring.ts
/**
 * knoll/scoring.ts — behavioral anomaly scoring engine (Phase 2).
 *
 * ADDITIVE to the six virtual laws (laws.ts) — it never replaces them. KNOLL runs the
 * laws first (structural / relational guarantees), then this scorer as a final
 * probabilistic gate for behavior the laws can't express as a hard rule (flooding,
 * high-entropy exfil blobs, priority abuse, bad source reputation over time).
 *
 * The scorer is stateful: it tracks a per-source sliding window of request timestamps
 * (rate) and an accumulating reputation risk (sources that keep tripping the scorer get
 * more suspicious). It remains monitor-only — it returns a verdict; it mutates no packet.
 *
 * ENFORCEMENT THRESHOLD (KNOLL active-router change): the default deny threshold is 0.34
 * (34%). This is intentionally aggressive — KNOLL is an INDEPENDENT ACTIVE ROUTER, not a
 * passive observer. Any packet whose behavioral anomaly score reaches 34% is denied AND
 * escalated by KNOLL into an immediate system-level FREEZE + packet QUARANTINE (see
 * knoll/freeze.ts, wired in knoll/validator.ts). The Shannon-entropy feature
 * (features.ts `intentEntropy`) is one of the contributors that can push a high-entropy
 * exfil-shaped blob across this 34% line.
 */
import { AgentRole, type RoutingPacket } from '../config/routing_schema.js';
import { extractFeatures, type BehavioralFeatures, type ScoringContext } from './features.js';

export interface BehavioralScore {
  score: number;
  threshold: number;
  /** score >= threshold — the packet should be denied. */
  isAnomalous: boolean;
  /** score >= flagThreshold but < threshold — allowed, but worth logging. */
  flagged: boolean;
  features: BehavioralFeatures;
  contributions: Record<keyof BehavioralFeatures, number>;
}

export type FeatureWeights = Record<keyof BehavioralFeatures, number>;

export interface BehavioralScorerOptions {
  /**
   * Deny threshold. score >= threshold → anomalous (deny + immediate freeze/quarantine).
   * Default 0.34 (34%). Lowered from the legacy 0.6 when KNOLL became an independent active
   * router: at 34% KNOLL denies the packet and trips a system-level freeze.
   */
  threshold?: number;
  /** Flag (log-but-allow) threshold. Kept below the deny threshold. Default 0.2. */
  flagThreshold?: number;
  /** Per-feature weights; defaults sum to 1.0. */
  weights?: Partial<FeatureWeights>;
  /** Sliding rate window in ms. Default 1000. */
  rateWindowMs?: number;
  /** Normalization cap for the rate feature. Default 20. */
  rateSoftCap?: number;
  /** Injectable clock for deterministic tests. */
  now?: () => number;
}

// Weights sum to 1.0. `intentEntropy` was raised (0.10 → 0.20, pulling from the low-signal
// `rate` and `payloadSize` features) so a Shannon-entropy SPIKE — a high-entropy exfil-shaped
// blob — can meaningfully contribute to crossing the 34% deny threshold. `maliciousHits` stays
// the dominant hard signal.
const DEFAULT_WEIGHTS: FeatureWeights = {
  rate: 0.1,
  intentEntropy: 0.2,
  maliciousHits: 0.3,
  endpointRisk: 0.15,
  payloadSize: 0.05,
  priorityAbuse: 0.1,
  sourceReputation: 0.1,
};

export class BehavioralScorer {
  private readonly threshold: number;
  private readonly flagThreshold: number;
  private readonly weights: FeatureWeights;
  private readonly rateWindowMs: number;
  private readonly rateSoftCap: number;
  private readonly now: () => number;

  private readonly windows = new Map<AgentRole, number[]>();
  private readonly reputation = new Map<AgentRole, number>();

  constructor(options: BehavioralScorerOptions = {}) {
    // 0.34 (34%) deny threshold — KNOLL active-router enforcement (was 0.6). See file header.
    this.threshold = options.threshold ?? 0.34;
    this.flagThreshold = options.flagThreshold ?? 0.2;
    this.weights = { ...DEFAULT_WEIGHTS, ...options.weights };
    this.rateWindowMs = options.rateWindowMs ?? 1000;
    this.rateSoftCap = options.rateSoftCap ?? 20;
    this.now = options.now ?? Date.now;
  }

  /**
   * Score a packet. Updates the source's rate window and (on an anomalous verdict) its
   * reputation risk, so repeat offenders climb toward denial over time.
   */
  score(packet: RoutingPacket): BehavioralScore {
    const source = packet.header.source;
    const recentCount = this.observeRate(source);
    const reputationRisk = this.reputation.get(source) ?? 0;

    const ctx: ScoringContext = {
      recentCount,
      rateSoftCap: this.rateSoftCap,
      reputationRisk,
    };

    const features = extractFeatures(packet, ctx);
    const contributions = {} as Record<keyof BehavioralFeatures, number>;
    let score = 0;
    for (const key of Object.keys(this.weights) as (keyof BehavioralFeatures)[]) {
      const c = features[key] * this.weights[key];
      contributions[key] = round4(c);
      score += c;
    }
    score = clamp01(score);

    const isAnomalous = score >= this.threshold;
    const flagged = !isAnomalous && score >= this.flagThreshold;

    // Reputation feedback loop: anomalous sources get worse; flagged nudge upward;
    // clean traffic slowly decays reputation back toward zero (good behavior forgiven).
    if (isAnomalous) {
      this.bumpReputation(source, 0.25);
    } else if (flagged) {
      this.bumpReputation(source, 0.05);
    } else {
      this.decayReputation(source, 0.02);
    }

    return {
      score: round4(score),
      threshold: this.threshold,
      isAnomalous,
      flagged,
      features,
      contributions,
    };
  }

  /** Read a source's current reputation risk (0 clean .. 1 bad). */
  reputationOf(source: AgentRole): number {
    return round4(this.reputation.get(source) ?? 0);
  }

  reset(): void {
    this.windows.clear();
    this.reputation.clear();
  }

  private observeRate(source: AgentRole): number {
    const t = this.now();
    const arr = this.windows.get(source) ?? [];
    const cutoff = t - this.rateWindowMs;
    const pruned = arr.filter((ts) => ts > cutoff);
    pruned.push(t);
    this.windows.set(source, pruned);
    return pruned.length;
  }

  private bumpReputation(source: AgentRole, delta: number): void {
    this.reputation.set(source, clamp01((this.reputation.get(source) ?? 0) + delta));
  }

  private decayReputation(source: AgentRole, delta: number): void {
    this.reputation.set(source, clamp01((this.reputation.get(source) ?? 0) - delta));
  }
}

function clamp01(n: number): number {
  return Math.max(0, Math.min(1, n));
}

function round4(n: number): number {
  return Math.round((n + Number.EPSILON) * 1e4) / 1e4;
}
